### PR TITLE
[border-router] adding `InfraIf` class

### PR DESCRIPTION
**Pull request by abtink**
This commit adds `InfraIf` class which represents an infrastructure
network interface on a border router providing methods & definitions
mirroring the platform APIs `otPlatInfraIf{}`. This commit also
updates the `RoutingManager` to to use the `InfraIf` class.

### CODE REVIEW

**Review by wgtdkp — APPROVED** [review]
Looks nice! 👍

File: src/core/border_router/infra_if.hpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+     *
+     * @param[in]  aInstance  A OpenThread instance.
+     *
+     */
+    explicit InfraIf(Instance &aInstance);
+
+    /**
+     * This method initializes the `InfraIf`.
+     *
+     * @param[in]  aInfraIfIndex    The infrastructure interface index.
```
Comment (jwhui) [inline src/core/border_router/infra_if.hpp:76]:
```suggestion
     * @param[in]  aIfIndex    The infrastructure interface index.
```

File: src/core/border_router/infra_if.hpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+     * This method deinitilaizes the `InfraIf`.
+     *
+     */
+    void Deinit(void);
+
+    /**
+     * This method indicates whether or not the `InfraIf` is initialized.
+     *
+     * @retval TRUE  The `InfraIf` is initialized.
+     * @retval TRUE  The `InfraIf` is not initialized.
```
Comment (jwhui) [inline src/core/border_router/infra_if.hpp:95]:
```suggestion
     * @retval TRUE   The `InfraIf` is initialized.
     * @retval FALSE  The `InfraIf` is not initialized.
```

File: src/core/border_router/infra_if.hpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+     *
+     */
+    bool IsInitialized(void) const { return mInitialized; }
+
+    /**
+     * This method indicates whether or not the infra interface is running.
+     *
+     * @retval TRUE  The infrastructure interface is running.
+     * @retval TRUE  The infrastructure interface is not running.
```
Comment (jwhui) [inline src/core/border_router/infra_if.hpp:104]:
```suggestion
     * @retval TRUE   The infrastructure interface is running.
     * @retval FALSE  The infrastructure interface is not running.
```

File: src/core/border_router/infra_if.hpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+     */
+    bool HasAddress(const Ip6::Address &aAddress);
+
+    /**
+     * This method sends an ICMPv6 Neighbor Discovery packet on the infrastructure interface.
+     *
+     * This method MUST be used when interface is initialized.
+     *
+     * @param[in]  aPacket        The ICMPv6 packet to send
+     * @param[in]  aDestAddress   The destination address.
```
Comment (jwhui) [inline src/core/border_router/infra_if.hpp:136]:
```suggestion
     * @param[in]  aPacket        The ICMPv6 packet to send.
     * @param[in]  aDestination   The destination address.
```

File: src/core/border_router/infra_if.hpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+ * This class represents the infrastructure network interface on a border router.
+ *
+ */
+class InfraIf : public InstanceLocator
+{
+public:
+    static constexpr uint16_t kInfoStringSize = 20; ///< Max chars for the info string (`ToString()`).
+
+    typedef String<kInfoStringSize> InfoString;  ///< String type returned from `ToString()`.
+    typedef Data<kWithUint16Length> Icmp6Packet; ///< An IMCP6 packet (data containing the IP payload)
```
Comment (jwhui) [inline src/core/border_router/infra_if.hpp:63]:
```suggestion
    typedef Data<kWithUint16Length> Icmp6Packet; ///< An ICMPv6 packet (data containing the IP payload).
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**